RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

```diff
@@ -0,0 +1,323 @@
+            (torch.nn.BatchNorm2d, torch.nn.Conv2d): torch.nn.intrinsic.qat.ConvBn2d
+        }
+    }
+}
+
+conv_functional_config = {
+“type”: QuantizedOperatorType.NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS,
+}
+
+bmm_config = {
+    “type”: QuantizedOperatorType.NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS,
+“fusions”: [(torch.nn.Softmax, torch.bmm)] # quantized_bmm_softmax
+}
+
+custom_backend_config_dict = {
+    # optional
+    "name": "custom_backend",
+    # quantized operator config is a map from
+    # module/functional/torch ops to their configurations
+    “operator”: {
+        torch.nn.Conv2d: conv_module_config,
+        torch.nn.functional.conv2d: conv_functional_config,
+        torch.bmm: bmm_config
+    }
+}
+
+# define a function to return the backend config dict
+def get_custom_backend_config_dict():
+    return custom_backend_config_dict
+
+# We'll also provide utility functions to get the backend configurations for a given backend:
+my_backend_config_dict = get_my_backend_config_dict()
+```
+
+# Proposed APIs for Model Developers
+```python
+from torch.quantization.quantize_fx import prepare_fx, convert_fx
+from custom_backend_library import get_my_backend_config_dict, lower_to_custom_backend
+
+backend_config_dict = get_my_backend_config_dict()
+model = prepare_fx(model, qconfig_dict, prepare_custom_config_dict = ?, backend_config_dict = backend_config_dict)
+# calibration
+...
+model = convert_fx(model, convert_custom_config_dict=?, backend_config_dict=backend_config_dict)
+
+# get the lower_to_custom_backend function defined by custom backend developers and call the function to lower a Reference Quantized Model to a model that runs on a custom backend
+model = lower_to_custom_backend(model)
+```
+
+# Use Cases
+## Use Case 1: Quantizing a Model for Inference on Server/Mobile
+```python
+from torch.quantization.quantize_fx import prepare_fx
+model = model.eval()
+qconfig_dict = {"": torch.quantization.default_qconfig}
+model = prepare_fx(model, qconfig_dict)
+calibration(model, ...)
+model = convert_fx(model)
+```
+
+The model produced here is a reference model that contains reference patterns, it is runnable since it is using quantize_per_tensor/dequantize/floating point operators to simulate quantized operators. For numerics we’ll provide an approximation to backend numerics even though it may not have the exact same numerics as any backends, or same speed up as the backends.
+
+### Backend Lowering (fbgemm/qnnpack)
+```python
+from torch.quantization.quantize_fx import prepare_fx
+model = model.eval()
+qconfig_dict = {"": torch.quantization.default_qconfig}
+model = prepare_fx(model, qconfig_dict)
+calibration(model, ...)
+model = convert_fx(model)
+
+# This step will transform a model with reference patterns to a model with
+# fbgemm/qnnpack ops, e.g. torch.ops.quantized.conv2d
+model = lower_to_fbgemm_fx(model)  # lower_to_qnnpack_fx(model)
+```
+
+
+### Example Implementation lower_to_fbgemm_fx
```

Comment:
the implementation of lower_to_fbgemm_fx is missing

Reply:
oh sorry, should add a `lower_to_fbgemm_fx` that calls `_lower_to_native_backend` below